llvm/lib/Analysis/DependenceAnalysis.cpp

```diff
@@ -3116,11 +3116,9 @@ const SCEV *DependenceInfo::addToCoefficient(const SCEV *Expr,
                                              const Loop *TargetLoop,
                                              const SCEV *Value) const {
   const SCEVAddRecExpr *AddRec = dyn_cast<SCEVAddRecExpr>(Expr);
-  if (!AddRec) // create a new addRec
-    return SE->getAddRecExpr(Expr,
-                             Value,
-                             TargetLoop,
-                             SCEV::FlagAnyWrap); // Worst case, with no info.
+  if (!AddRec)
+    return SE->getAddRecExpr(Expr, Value, TargetLoop, SCEV::FlagNSW);
```

Comment:
Why is it safe to assume NSW without any additional information here? 

Also, I think as @nikic mentioned, at the moment, we would need to prove that NSW holds in all contexts where the new AddRec is valid

There's https://github.com/llvm/llvm-project/pull/91961 to extend SCEV to allow constructing SCEV's with context-sensitive wrapping flags.

Reply:
> Why is it safe to assume NSW without any additional information here?

Here's the additional info here.
The AddRec created here is only used in the dependence test.  It does not represent the evolution of a scalar variable from the compiled code.  The AddRec is used to represent the remaining dependence relation after successive projections.  The intended semantics is that of a linear function (i.e., a function that does not wrap around.)  Adding wrap-around semantics to the SCEV makes the relation non-linear.  classifyPair then `return Subscript::NonLinear;` and this breaks the dependence test as the non-linear access comes as a surprise this late in the test: all non-linear memory access functions have been filtered out by this time.

> we would need to prove that NSW holds in all contexts where the new AddRec is valid

The AddRec created in here is only valid in the data dependence test.
The AddRec is used as a representation of the dependence relation.